File: app/lib/shared/models/flight.dart
import 'airport.dart';
import 'airplane.dart';
import 'enums.dart';

class Flight {
  final int id;
  final String flightNumber;
  final int airplaneId;
  final int originAirportId;
  final int destinationAirportId;
  final DateTime departureTime;
  final DateTime arrivalTime;
  final double basePrice;
  final String? gate;
  final String? terminal;
  final FlightStatus status;
  final DateTime? boardingTime;
  final Airport? origin;
  final Airport? destination;
  final Airplane? airplane;
  final int? durationMinutes;
  final int? availableSeats;

  Flight({
    required this.id,
    required this.flightNumber,
    required this.airplaneId,
    required this.originAirportId,
    required this.destinationAirportId,
    required this.departureTime,
    required this.arrivalTime,
    required this.basePrice,
    this.gate,
    this.terminal,
    required this.status,
    this.boardingTime,
    this.origin,
    this.destination,
    this.airplane,
    this.durationMinutes,
    this.availableSeats,
  });

  factory Flight.fromJson(Map<String, dynamic> json) {
    return Flight(
      id: json['id'] != null ? (json['id'] as num).toInt() : 0,
      flightNumber: json['flight_number'] as String,
      airplaneId: json['airplane_id'] != null ? (json['airplane_id'] as num).toInt() : 0,
      originAirportId: json['origin_airport_id'] != null ? (json['origin_airport_id'] as num).toInt() : 0,
      destinationAirportId: json['destination_airport_id'] != null ? (json['destination_airport_id'] as num).toInt() : 0,
      departureTime: DateTime.parse(json['departure_time'] as String),
      arrivalTime: DateTime.parse(json['arrival_time'] as String),
      basePrice: json['base_price'] != null ? (json['base_price'] as num).toDouble() : 0.0,
      gate: json['gate'] as String?,
      terminal: json['terminal'] as String?,
      status: FlightStatus.fromJson(json['status'] as String),
      boardingTime: json['boarding_time'] != null
          ? DateTime.parse(json['boarding_time'] as String)
          : null,
      origin: json['origin_airport'] != null
          ? Airport.fromJson(json['origin_airport'] as Map<String, dynamic>)
          : null,
      destination: json['destination_airport'] != null
          ? Airport.fromJson(json['destination_airport'] as Map<String, dynamic>)
          : null,
      airplane: json['airplane'] != null
          ? Airplane.fromJson(json['airplane'] as Map<String, dynamic>)
          : null,
      durationMinutes: json['duration_minutes'] != null 
          ? (json['duration_minutes'] as num).toInt() 
          : null,
      availableSeats: json['available_seats'] != null 
          ? (json['available_seats'] as num).toInt() 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'flight_number': flightNumber,
      'airplane_id': airplaneId,
      'origin_airport_id': originAirportId,
      'destination_airport_id': destinationAirportId,
      'departure_time': departureTime.toIso8601String(),
      'arrival_time': arrivalTime.toIso8601String(),
      'base_price': basePrice,
      'gate': gate,
      'terminal': terminal,
      'status': status.toJson(),
      'boarding_time': boardingTime?.toIso8601String(),
      if (origin != null) 'origin': origin!.toJson(),
      if (destination != null) 'destination': destination!.toJson(),
      if (airplane != null) 'airplane': airplane!.toJson(),
      if (durationMinutes != null) 'duration_minutes': durationMinutes,
      if (availableSeats != null) 'available_seats': availableSeats,
    };
  }
  
  Duration get duration => arrivalTime.difference(departureTime);
}

File: AirlineAIT/backend/app/schemas/flight.py
from pydantic import BaseModel, field_validator
from datetime import datetime
from typing import Optional, List, Dict, Any
from app.models.flight import FlightStatus
from app.schemas.airport import AirportResponse


class FlightSearch(BaseModel):
    origin_airport_id: Optional[int] = None
    destination_airport_id: Optional[int] = None
    departure_date: Optional[datetime] = None


class FlightResponse(BaseModel):
    id: int
    flight_number: str
    origin_airport_id: int
    destination_airport_id: int
    airplane_id: int
    departure_time: datetime
    arrival_time: datetime
    price: float
    category_prices: Optional[Dict[str, float]] = None
    category_allocations: Optional[Dict[str, Dict[str, int]]] = None
    gate: Optional[str]
    terminal: Optional[str]
    status: str
    origin_airport: Optional[AirportResponse] = None
    destination_airport: Optional[AirportResponse] = None
    duration_minutes: Optional[int] = None  # Calculated duration in minutes
    
    class Config:
        from_attributes = True
    
    @classmethod
    def from_orm_with_duration(cls, flight):
        """Create response with calculated duration"""
        duration = None
        if flight.departure_time and flight.arrival_time:
            duration = int((flight.arrival_time - flight.departure_time).total_seconds() / 60)
        data = cls.model_validate(flight).model_dump()
        data['duration_minutes'] = duration
        return cls(**data)
    
    @field_validator('category_prices', 'category_allocations', mode='before')
    @classmethod
    def parse_json_fields(cls, v):
        if isinstance(v, str):
            import json
            try:
                return json.loads(v)
            except json.JSONDecodeError:
                return None
        return v


class FlightDetailResponse(FlightResponse):
    airplane: Dict[str, Any]
    available_seats: int


class SeatMapResponse(BaseModel):
    flight_id: int
    seat_map: Dict[str, Any]  # {seat_number: {available: bool, category: str, held_until: Optional[datetime]}}
    template: Dict[str, Any]


class FlightCreate(BaseModel):
    flight_number: str
    origin_airport_id: int
    destination_airport_id: int
    airplane_id: int
    departure_time: datetime
    arrival_time: datetime
    price: float
    category_prices: Optional[Dict[str, float]] = None
    category_allocations: Optional[Dict[str, Dict[str, int]]] = None
    gate: Optional[str] = None
    terminal: Optional[str] = None
    status: FlightStatus = FlightStatus.SCHEDULED


class FlightUpdate(BaseModel):
    airplane_id: Optional[int] = None
    departure_time: Optional[datetime] = None
    arrival_time: Optional[datetime] = None
    price: Optional[float] = None
    category_prices: Optional[Dict[str, float]] = None
    category_allocations: Optional[Dict[str, Dict[str, int]]] = None
    gate: Optional[str] = None
    terminal: Optional[str] = None
    status: Optional[FlightStatus] = None

